### PR TITLE
Add message type and list API docs

### PR DESCRIPTION
**Pull request by jugarrit**
This PR brings with it all of the changes from adding the `type` fields to postMessage, as well as the new List API documentation.

### CODE REVIEW

File: source/includes/rest/_core.md [hunk 1]
```diff
@@ -1308,9 +1310,10 @@ smooch.appUsers.sendMessage('c7f6e6d6c3a637261bd9656f', {
 
 <api>`POST /v1/appusers/{smoochId|userId}/messages`</api>
 
-Post a message to the app user. If the app user does not yet have a conversation, one will be created automatically. Messages must have a `role` of either `appuser` or `appMaker`. Furthermore, a message must have a `text`, `actions`, `items`, or `mediaUrl` specified. Some parameters are mutually exclusive, indicated in the table blow.
+Post a message to the app user. If the app user does not yet have a conversation, one will be created automatically. Messages must have a `role` of either `appuser` or `appMaker`. Furthermore, a message must have a `type` specifying the type of message you're trying to send.
```
Comment (mspensieri) [inline source/includes/rest/_core.md]: > Post a message to the app user

to or from...

File: source/includes/rest/_core.md [hunk 2]
```diff
@@ -1319,250 +1322,159 @@ For messages originating from an app maker, a `jwt` credential with `app` level
...
+201 CREATED
+```
+```json
 {
-    "text":"Which do you prefer?",
+  "message": {
+    "_id": "57966d21c19c9da00839a5e9",
     "role": "appMaker",
+    "type": "text",
+    "mediaUrl": "http://example.org/image.jpg",
```
Comment (mspensieri) [inline source/includes/rest/_core.md]: Missing `mediaType` in this response

Comment (mspensieri) [inline source/includes/rest/_core.md]: Same thing for the carousel items below, missing the media type

Comment (mspensieri) [inline source/includes/rest/_core.md]: List items too!

File: source/includes/rest/_core.md [hunk 3]
```diff
@@ -1688,38 +1603,535 @@ smooch.appUsers.sendMessage('c7f6e6d6c3a637261bd9656f', {
...
+
+<aside class="notice">
+    In Messenger, list messages sent with multiple `actions` either in a `message item` or the `message` itself will be truncated, as only 1 action is supported.
+</aside>
+
+<span class="half-width-img">![messenger list](/images/list_messenger.png)</span>
+
+## Message Items
+
+Message items can be sent through the [post message](#post-message) by including them in the message payload.
```
Comment (mspensieri) [inline source/includes/rest/_core.md]: "through the post message" -> through the post message API

File: source/includes/rest/_core.md [hunk 4]
```diff
@@ -1688,38 +1603,535 @@ smooch.appUsers.sendMessage('c7f6e6d6c3a637261bd9656f', {
...
+|------------------------------|------------------------|
+| **title**<br/><span class='req'>required</span>       | The title of the carousel item. |
+| **actions**<br/><span class='req'>required</span>     | Array of [action buttons](#action-buttons). At least 1 is required, a maximum of 3 are allowed. `link` and `postback` and `share` actions are supported. |
+| **description**<br/><span class='opt'>optional</span> | The text description, or subtitle. |
+| **mediaUrl**<br/><span class='opt'>optional</span>    | The image URL to be shown in the carousel/list item. |
+| **size**<br/><span class='opt'>optional</span>        | The size of the image to be shown in the carousel/list item (Only top item of Facebook Messenger carousel supportted). Choose from `compact` and `large` |
+| **mediaType**<br/><span class='opt'>optional</span>   | If a `mediaUrl` was specified, the media type is defined here, for example `image/jpeg`. If `mediaType` is not specified, the media type will be resolved with the `mediaUrl`. |
+
+## Action Buttons
+Actions buttons can be sent through the [post message](#post-message) by including them in the message payload.
```
Comment (mspensieri) [inline source/includes/rest/_core.md]: same comment, post message API

File: source/includes/rest/_core.md (hunk 2 above)
Comment (jugarrit) [inline source/includes/rest/_core.md]: Wasn't there in the original payload so I wasn't sure if it'd be present!

File: source/includes/rest/_core.md [hunk 5]
```diff
@@ -1688,38 +1603,535 @@ smooch.appUsers.sendMessage('c7f6e6d6c3a637261bd9656f', {
...
+```
+
+> Response:
+
+```
+201 CREATED
+```
+```json
+{
+  "message": {
```
Comment (jugarrit) [inline source/includes/rest/_core.md]: non-matching payload will be fixed in next push

File: source/includes/rest/_core.md [hunk 6]
```diff
@@ -1308,9 +1310,10 @@ smooch.appUsers.sendMessage('c7f6e6d6c3a637261bd9656f', {
 
 <api>`POST /v1/appusers/{smoochId|userId}/messages`</api>
 
-Post a message to the app user. If the app user does not yet have a conversation, one will be created automatically. Messages must have a `role` of either `appuser` or `appMaker`. Furthermore, a message must have a `text`, `actions`, `items`, or `mediaUrl` specified. Some parameters are mutually exclusive, indicated in the table blow.
+Post a message to or from the app user. If the app user does not yet have a conversation, one will be created automatically. Messages must have a `role` of either `appuser` or `appMaker`. Furthermore, a message must have a `type` specifying the type of message you're trying to send.
```
Comment (jpjoyal) [inline source/includes/rest/_core.md]: Must is pretty strong no ? Its still optional in v1 for sure

File: source/includes/rest/_core.md [hunk 7]
```diff
@@ -1308,9 +1310,10 @@ smooch.appUsers.sendMessage('c7f6e6d6c3a637261bd9656f', {
...
+Depending on the type, the message object will have additional `Required` and `Optional` arguments.
 
-Images can be posted by reference using this API by specifying the `mediaUrl` parameter. Alternatively, you may also upload images to the conversation directly using the [`/images`](#upload-image) endpoint.
+Images can be posted by reference using this API using the `image` type. Alternatively, you may also upload images to the conversation directly using the [`/images`](#upload-image) endpoint.
```
Comment (jpjoyal) [inline source/includes/rest/_core.md]: by reference is a little weird no ? by URL would be better imo

File: source/includes/rest/_core.md [hunk 8]
```diff
@@ -1308,9 +1310,10 @@ smooch.appUsers.sendMessage('c7f6e6d6c3a637261bd9656f', {
 
 <api>`POST /v1/appusers/{smoochId|userId}/messages`</api>
 
-Post a message to the app user. If the app user does not yet have a conversation, one will be created automatically. Messages must have a `role` of either `appuser` or `appMaker`. Furthermore, a message must have a `text`, `actions`, `items`, or `mediaUrl` specified. Some parameters are mutually exclusive, indicated in the table blow.
+Post a message to or from the app user. If the app user does not yet have a conversation, one will be created automatically. Messages must have a `role` of either `appuser` or `appMaker`. Furthermore, a message must have a `type` specifying the type of message you're trying to send.
+Depending on the type, the message object will have additional `Required` and `Optional` arguments.
```
Comment (jpjoyal) [inline source/includes/rest/_core.md]: Maybe add: "see text, image, carousel, list sections" with links. Might actually be worth to separate the type discussion into his own paragraph

File: source/includes/rest/_core.md [hunk 9]
```diff
@@ -1319,250 +1322,160 @@ For messages originating from an app maker, a `jwt` credential with `app` level
...
-## Action buttons
-Actions buttons can be sent through the [post message](#post-message) by including them in the message payload.
-
-There are 4 types of supported actions : **link**, **buy**, **postback**, and **reply**. Type must be specified by providing a `type` argument in the action object.
-
 <aside class="notice">
-    Action buttons can only be sent with an `appMaker` role.
+Additional arguments necessary based on message type ([`text`](#text-message), [`image`](#image-message), [`carousel`](#carousel-message), [`list`](#list-message))
+<br/>
+For backwards compatibility reasons, we will try to infer a message `type` if one is not provided.
```
Comment (jpjoyal) [inline source/includes/rest/_core.md]: I would phrases as `type` is inferred when not provided.

File: source/includes/rest/_core.md [hunk 10]
```diff
@@ -1247,35 +1247,37 @@ Typing activity is only supported on our Web Messenger, Facebook Messenger and T
...
 ```
 
-> Post as app maker:
+> Request (App Maker):
 
 ```shell
 curl https://api.smooch.io/v1/appusers/c7f6e6d6c3a637261bd9656f/messages \
      -X POST \
-     -d '{"text":"Just put some vinegar on it", "role": "appMaker"}' \
+     -d '{"text":"Just put some vinegar on it", "role": "appMaker", "type": "text"}' \
```
Comment (jugarrit) [inline source/includes/rest/_core.md]: We can't make it mandatory until V2, but the idea is that it's technically mandatory from this point on.

File: source/includes/rest/_core.md (hunk 7 above)
Comment (jugarrit) [inline source/includes/rest/_core.md]: Sure!

File: source/includes/rest/_core.md [hunk 11]
```diff
@@ -1319,250 +1322,160 @@ For messages originating from an app maker, a `jwt` credential with `app` level
...
 | **payload**<br/><span class='opt'>optional</span>    | The payload of a `reply` action, if applicable |
 
-## Action buttons
-Actions buttons can be sent through the [post message](#post-message) by including them in the message payload.
-
-There are 4 types of supported actions : **link**, **buy**, **postback**, and **reply**. Type must be specified by providing a `type` argument in the action object.
-
 <aside class="notice">
-    Action buttons can only be sent with an `appMaker` role.
+Additional arguments necessary based on message type ([`text`](#text-message), [`image`](#image-message), [`carousel`](#carousel-message), [`list`](#list-message))
```
Comment (jugarrit) [inline source/includes/rest/_core.md]: What do you mean by "we should be explicit"? 

Arguments necessary is still grammatically correct, albeit it's a little robotic sounding.
Will fix.

File: source/includes/rest/_core.md (hunk 8 above)
Comment (jugarrit) [inline source/includes/rest/_core.md]: We also have the notice just below. Worth having it here too?

File: source/includes/rest/_core.md [hunk 12]
```diff
@@ -1308,9 +1310,13 @@ smooch.appUsers.sendMessage('c7f6e6d6c3a637261bd9656f', {
 
 <api>`POST /v1/appusers/{smoochId|userId}/messages`</api>
 
-Post a message to the app user. If the app user does not yet have a conversation, one will be created automatically. Messages must have a `role` of either `appuser` or `appMaker`. Furthermore, a message must have a `text`, `actions`, `items`, or `mediaUrl` specified. Some parameters are mutually exclusive, indicated in the table blow.
+Post a message to or from the app user. If the app user does not yet have a conversation, one will be created automatically. Messages must have a `role` of either `appuser` or `appMaker`.
```
Comment (mspensieri) [inline source/includes/rest/_core.md]: hmm missed this, incorrect capitalization in appuser role -> `appUser`

Comment (jugarrit) [inline source/includes/rest/_core.md]: Good catch.

File: source/includes/rest/_core.md [hunk 13]
```diff
@@ -1319,250 +1325,160 @@ For messages originating from an app maker, a `jwt` credential with `app` level
...
-## Action buttons
-Actions buttons can be sent through the [post message](#post-message) by including them in the message payload.
-
-There are 4 types of supported actions : **link**, **buy**, **postback**, and **reply**. Type must be specified by providing a `type` argument in the action object.
-
 <aside class="notice">
-    Action buttons can only be sent with an `appMaker` role.
+Additional arguments are necessary based on message type ([`text`](#text-message), [`image`](#image-message), [`carousel`](#carousel-message), [`list`](#list-message))
+<br/>
+For backwards compatibility reasons, `type` is inferred when not provided.
```
Comment (alavers) [inline source/includes/rest/_core.md]: I'd suggest we just not mention this at all

File: source/includes/rest/_core.md [hunk 14]
```diff
@@ -1688,38 +1609,590 @@ smooch.appUsers.sendMessage('c7f6e6d6c3a637261bd9656f', {
...
+> Request:
 
 ```shell
 curl https://api.smooch.io/v1/appusers/c7f6e6d6c3a637261bd9656f/messages \
      -X POST \
      -H 'content-type: application/json' \
-     -H "authorization: Bearer your-jwt" \
+     -H 'authorization: Bearer your-jwt' \
      -d '
+     {
```
Comment (alavers) [inline source/includes/rest/_core.md]: Sample payload probably doesn't need to be this long :] 1-2 items showing nested actions is probably sufficient I think

**Pull request merged**